engine/sync_manager.go

```diff
@@ -321,27 +308,55 @@ func (m *syncManager) add(c *currencyPairSyncAgent) {
 		}
 	}
 
-	c.Created = time.Now()
-	m.currencyPairs = append(m.currencyPairs, *c)
+	if m.currencyPairs == nil {
+		m.currencyPairs = make(map[string]map[*currency.Item]map[*currency.Item]map[asset.Item]*currencyPairSyncAgent)
+	}
+
+	m1, ok := m.currencyPairs[c.Exchange]
+	if !ok {
+		m1 = make(map[*currency.Item]map[*currency.Item]map[asset.Item]*currencyPairSyncAgent)
+		m.currencyPairs[c.Exchange] = m1
+	}
+
+	m2, ok := m1[c.Pair.Base.Item]
+	if !ok {
+		m2 = make(map[*currency.Item]map[asset.Item]*currencyPairSyncAgent)
+		m1[c.Pair.Base.Item] = m2
+	}
+
+	m3, ok := m2[c.Pair.Quote.Item]
+	if !ok {
+		m3 = make(map[asset.Item]*currencyPairSyncAgent)
+		m2[c.Pair.Quote.Item] = m3
+	}
+
+	_, ok = m3[c.AssetType]
+	if ok {
+		log.Errorf(log.SyncMgr, "%s %s %s %v", c.Exchange, c.Pair, c.AssetType, errSyncItemAlreadyAdded)
+		return
+	}
+
+	cpy := *c
+	cpy.Created = time.Now()
+	m3[c.AssetType] = &cpy
```

Comment:
Yeah, I have a rule of thumb `now` 😆, if a caller passes in a pointer into a function and it's going to store it, might as well make a copy because I don't know what the caller is going to do with it.